### PR TITLE
docker_swarm update to use shared Swarm library

### PR DESCRIPTION
**Pull request by WojciechowskiPiotr**
##### SUMMARY
Update the `docker_swarm` module so it use the `ansible.module_utils.docker.swarm` class `AnsibleDockerSwarmClient` for client connectivity as the rest of the Swarm modules. It also updates the library for this module specific call of method `check_if_swarm_node_is_down` and adds deprecation warning for `state: inspect`

There is no new feature nor change in usage so changelog update is not required

Discussed in ansible/community#408

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_swarm

### CODE REVIEW

**Comment by WojciechowskiPiotr** [comment]
ready_for_review

**Review by felixfontein — CHANGES_REQUESTED** [review]
Looks good. Since this touches a lot of code and can't be backported, I don't want to merge this before all important 2.7 problems with docker_swarm are fixed.

(Marking this as "request changes" to prevent accidental merging.)

**Comment by felixfontein** [comment]
I just looked again through the issues, I think the most pressing/important one is #51175, which should be fixed by #52895 (that also includes some refactoring to fix problems with idempotency, and almost includes a proper `--diff` mode -- I'll add a Feature PR for that once this is merged). Can you take a look at that PR?

**Comment by WojciechowskiPiotr** [comment]
Maybe it will be easier to split it into smaller PRs? At least should be a good idea to move to swarm library because otherwise, if you fix something in other PR there is no way to test your patch to check if other modules using the library are still working

**Comment by felixfontein** [comment]
The changes from the other PR should not affect anything for the swarm library. They only fix problems with the docker_swarm module.

**Comment by felixfontein** [comment]
#52895 has been merged, so now this one can proceed once you rebase it.

File: lib/ansible/modules/cloud/docker/docker_swarm.py [hunk 1]
```diff
@@ -158,6 +158,7 @@
   - "docker >= 2.6.0"
   - Docker API >= 1.25
 author:
+  - Piotr Wojciechowski (@WojciechowskiPiotr)
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: I think you should add your name at the end of the list, not at the beginning.

File: lib/ansible/modules/cloud/docker/docker_swarm.py [hunk 2]
```diff
@@ -406,7 +404,8 @@ def inspect_swarm(self):
             return
 
     def init_swarm(self):
-        if self.__isSwarmManager():
+        if self.client.check_if_swarm_manager():
+            self.inspect_swarm()
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: That's not needed, `__update_swarm()` already calls `inspect_swarm()`.

File: lib/ansible/modules/cloud/docker/docker_swarm.py [hunk 3]
```diff
@@ -386,16 +390,10 @@ def __call__(self):
...
-        try:
-            data = self.client.inspect_swarm()
-            json_str = json.dumps(data, ensure_ascii=False)
-            self.swarm_info = json.loads(json_str)
-            return True
-        except APIError:
-            return False
-
     def inspect_swarm(self):
+        self.client.module.deprecate(
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm.py]: This is now called every time the swarm is updated. It should only be called if `state` is `inspect`.

**Review by felixfontein — COMMENTED** [review]
For some reason, the idempotency checks fail. I have to investigate more to find out why.

**Comment by felixfontein** [comment]
I think I found the reason for the idempotency problems; I've created a PR for that: #52976

**Comment by WojciechowskiPiotr** [comment]
> For some reason, the idempotency checks fail. I have to investigate more to find out why.

The reason is that `init_swarm()` if run on a host that is already part of swarm cluster calls `__update_swarm()` which changes the `Version.Index` when updating it. Lines 407-410.

**Comment by felixfontein** [comment]
> The reason is that `init_swarm()` if run on a host that is already part of swarm cluster calls `__update_swarm()` which changes the `Version.Index` when updating it. Lines 407-410.

Not really: `Version.Index` is not used to determine whether the swarm should be updated. That only depends on some of the fields in `TaskParameters`. (I cloned your branch and added the diff mode from #52947 to find out why it actually wanted to update the swarm.)

File: lib/ansible/modules/cloud/docker/docker_swarm.py (hunk 3 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm.py]: Adding a `if self.state == 'inspect': ` to `def __call__()` and putting the deprecation warning there (before the dispatching is done) is probably the cleanest solution.

File: lib/ansible/modules/cloud/docker/docker_swarm.py [hunk 4]
```diff
@@ -391,20 +395,16 @@ def __call__(self):
 
         choice_map.get(self.state)()
 
+        if self.state == 'inspect':
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm.py:396]: Can you move this two lines up, so that the dispatching comes after the deprecation warning?

Comment (WojciechowskiPiotr) [reply lib/ansible/modules/cloud/docker/docker_swarm.py:396]: No problem :)

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm.py:396]: Thanks! I think it's better to issue the warning before "real action" starts (so the errors don't block the warning).

**Comment by felixfontein** [comment]
Hmm, is it just me, or are these read timeouts becoming more common? https://app.shippable.com/github/ansible/ansible/runs/110389/57/tests https://app.shippable.com/github/ansible/ansible/runs/110389/58/tests

**Comment by felixfontein** [comment]
(I've also seen that with all other docker_swarm PRs today, so it's not specific to this PR.)

**Comment by felixfontein** [comment]
Instead of an empty commit, you can also simply close and re-open the PR; ansibot will then trigger a new CI run.

**Comment by felixfontein** [comment]
shipit

**Comment by felixfontein** [comment]
@WojciechowskiPiotr thanks for this PR!

**Comment by WojciechowskiPiotr** [comment]
No problem

File: lib/ansible/modules/cloud/docker/docker_swarm.py [hunk 5]
```diff
@@ -253,6 +256,7 @@ def __init__(self):
         self.force_new_cluster = None
         self.remote_addrs = None
         self.join_token = None
+        self.force = None
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm.py:259]: Ah, this one should not have been there. I guess this is a rebasing artefact. I'll create a PR to get rid of it.

**Pull request merged**